sdk/messaging/azeventgrid/publisher/client_custom.go

```diff
@@ -0,0 +1,126 @@
+//go:build go1.18
+// +build go1.18
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+package publisher
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/messaging"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azeventgrid/internal"
+)
+
+// ClientOptions contains optional settings for [Client]
+type ClientOptions struct {
+	azcore.ClientOptions
+}
+
+// NewClientWithSharedKeyCredential creates a [Client] using a shared key.
+func NewClientWithSAS(endpoint string, sas string, options *ClientOptions) (*Client, error) {
+	const sasTokenHeader = "aeg-sas-token"
+
+	if options == nil {
+		options = &ClientOptions{}
+	}
+
+	// TODO: I believe we're supposed to allow for dynamically updating the key at any time as well.
+	azc, err := azcore.NewClient(internal.ModuleName+".PublisherClient", internal.ModuleVersion, runtime.PipelineOptions{
+		PerRetry: []policy.Policy{
+			// TODO: Java has a specific policy for this kind of authentication.
+			// AzureKeyCredentialPolicy
+			&skpolicy{HeaderName: sasTokenHeader, Key: sas},
+		},
+	}, &options.ClientOptions)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		internal: azc,
+		endpoint: endpoint,
+	}, nil
+}
+
+// NewClientWithSharedKeyCredential creates a [Client] using a shared key.
+func NewClientWithSharedKeyCredential(endpoint string, key string, options *ClientOptions) (*Client, error) {
```

Comment:
`key` should probably be a `SharedKeyCredential` type and would fix your TODO below.  See what we did in `azblob` for an example.

Reply:
(per our offline discussion)

Let's delay this change and create an actual credential type for this in azcore (tracked with this issue: #20541). I'd want to use this same key type in the root package of this module as well, and there's even a few other places like azopenai.

(not sure if the ship has sailed for azblob, but I'm guessing we can work it out with a type alias).